ledger-service/lf-value-json/specification.rst

```diff
@@ -0,0 +1,434 @@
+    '2019-06-18T08:59:34.191Z'
+
+Python
+
+::
+
+    >>> datetime.datetime.utcnow().isoformat() + 'Z'
+    '2019-06-18T08:59:08.392764Z'
+
+Java
+
+::
+
+    import java.time.Instant;
+    class Main {
+        public static void main(String[] args) {
+            Instant instant = Instant.now();
+            // prints 2019-06-18T09:02:16.652Z
+            System.out.println(instant.toString());
+        }
+    }
+
+Output
+~~~~~~
+
+Timestamps are encoded as ISO 8601 strings, rendered using the format
+``yyyy-mm-ddThh:mm:ss[.ssssss]Z``.
+
+The sub-second part will be formatted as follows:
+
+- If no sub-second part is present in the timestamp (i.e. the timestamp
+  represents whole seconds), the sub-second part will be omitted
+  entirely;
+- If the sub-second part does not go beyond milliseconds, the sub-second
+  part will be up to milliseconds, padding with trailing 0s if
+  necessary;
+- Otherwise, the sub-second part will be up to microseconds, padding
+  with trailing 0s if necessary.
+
+In other words, the encoded timestamp will either have no sub-second
+part, a sub-second part of length 3, or a sub-second part of length 6.
+
+Party
+-----
+
+Represented using their string representation, without any additional
+quotes::
+
+    "Alice"
+    "Bob"
+
+Unit
+----
+
+Represented as ``{}``. Note that ``{} !== {}``, however, ``null`` would
+be ambiguous; for the type ``Optional Unit``, ``null`` decodes to
+``None``, but ``{}`` decodes to ``Some ()``.
+
+Additionally, we think that this is the least confusing encoding for
+Unit since unit is conceptually an empty record.  We do not want to
+imply that Unit is used similarly to null in JavaScript or None in
+Python.
+
+Date
+----
+
+Represented as an ISO 8601 date rendered using the format
+``yyyy-mm-dd``::
+
+    2019-06-18
+    9999-12-31
+    0001-01-01
+
+The dates must be between the bounds specified by DAML-LF and ISO 8601,
+[0001-01-01, 9999-99-99].
+
+Text
+----
+
+Represented as JSON strings.
+
+Bool
+----
+
+Represented as JSON booleans.
+
+Record
+------
+
+Input
+~~~~~
+
+Records can be represented in two ways. As JSON objects::
+
+    { f₁: v₁, ..., fₙ: vₙ }
+
+And as arrays::
+
+    [ v₁, ..., vₙ ]
+
+Note that DAML-LF record fields are ordered. So if we have
+
+::
+
+    record Foo = {f1: Int64, f2: Bool}
+
+when representing the record as an array the user must specify the
+fields in order::
+
+    [42, true]
+
+The motivation for the array format for records is to allow specifying
+tuple types closer to what it looks like in DAML. Note that a DAML
+tuple, i.e. (42, True), will be compiled to a DAML-LF record ``Tuple2 {
+_1 = 42, _2 = True }``.
+
+Output
+~~~~~~
+
+Records are always encoded as JSON objects.
+
+List
+----
+
+Lists are represented as
+
+::
+
+    [v₁, ..., vₙ]
+
+Map
+---
+
+Maps are represented as JSON objects:
+
+::
+
+    { k₁: v₁, ..., kₙ: vₙ }
+
+Optional
+--------
+
+Input
+~~~~~
+
+Optionals are encoded using ``null`` if the value is None, and with the
```

Comment:
if it is up to me, I would always encode DAML Optional as `JsArray`: [], just for consistency.
```
[]         --> None
[42]     --> Some 42
[[42]]   --> Some (Some 42)
```

Reply:
but at the same time this would require specifying empty `JsArray` for every optional field in the JSON command. Which is a bit too much, I guess. So, scratch the above.